jobqueue_features/tests/test_decorators.py

```diff
@@ -1,41 +1,316 @@
-from jobqueue_features.clusters_controller import clusters_controller_singleton
+from jobqueue_features.clusters import CustomSLURMCluster
+from jobqueue_features.clusters_controller import clusters_controller_singleton as controller
 
 
-class TestDecorators(TestCase):
-    def setUp(self):
-        self.cluster_type = LocalCluster
-        set_default_cluster(self.cluster_type)
+class TweakedCustomCluster(CustomSLURMCluster):
+    def __init__(self, **kwargs):
+        kwargs["interface"] = ""
+        super(TweakedCustomCluster, self).__init__(**kwargs)
+
+
+# monkey patch the CustomSLURMCluster so it can be initialized without providing
+# the "interface" attribute
+CustomSLURMCluster = TweakedCustomCluster
+
+
+class TestOnClusterDecorator(TestCase):
+    """
+    Test @on_cluster() decorator against several scenarios of providing clusters or clusters
+    IDs to the initializer of decorator for getting/adding clusters from cluster controller.
+
+    cluster=None,         cluster_id=None : default cluster
+    cluster=SLURMCluster, cluster_id=None : register cluster by its name
+    cluster=LocalCluster, cluster_id=None : error
+    cluster=None,         cluster_id=str  : when default cluster is LocalCLuster check for cluster with such id
+    cluster=None,         cluster_id=str  : when default cluster is SLURMCluster check for cluster with such id
+    cluster=SLURMCluster, cluster_id=str  : assert cluster.name == id
+    cluster=LocalCluster, cluster_id=str  : register cluster by given id
+    """
+    def tearDown(self):
+        controller._close()
+
+    def test_default_behavior(self):
+        self._test_default_behavior(cluster_type=CustomSLURMCluster)
+
+    def test_default_behavior_local(self):
+        self._test_default_behavior(cluster_type=LocalCluster)
+
+    def test_init_only_cluster(self):
+        original_cluster = CustomSLURMCluster()
+
+        @on_cluster(cluster=original_cluster)
+        def f():
+            pass
+
+        cluster, client = controller.get_cluster(original_cluster.name)
+        self.assertEqual(original_cluster, cluster)
+        self.assertIsInstance(client, Client)
+        controller.delete_cluster(cluster.name)
+
+        _id = "test1"
+        original_cluster = CustomSLURMCluster(name=_id)
+
+        @on_cluster(cluster=original_cluster)
+        def f():
+            pass
+
+        cluster, client = controller.get_cluster(_id)
+        self.assertEqual(original_cluster, cluster)
+        self.assertIsInstance(client, Client)
+        controller.delete_cluster(cluster.name)
+
+    def test_init_only_cluster_local(self):
+        original_cluster = LocalCluster()
+        with self.assertRaises(ClusterException) as ctx:
+            @on_cluster(cluster=original_cluster)
+            def f():
+                pass
+        self.assertEqual('LocalCluster requires "cluster_id" argument.',
+                         ctx.exception.message)
```

Comment:
From https://stackoverflow.com/a/43453668 you should probably use `str(ctx.exception)`

Reply:
Done